test_main.py

```diff
@@ -19,13 +22,26 @@ def message(self):
 
     return MockResponse()
 
+@pytest.fixture(autouse=True)
+def set_openai_api_key():
```

Comment:
Can you put these util methods into their own files?

Reply:
Certainly! We can move the utility methods into their own files to keep the test file clean and organized. I'll make the necessary changes and update the PR.

### Plan

1. **Create a New File for Test Utilities:**
   - Create a new file `test_utils.py` to hold the utility methods and fixtures.

2. **Move Utility Methods to `test_utils.py`:**
   - Move the `mock_openai_chat_completion_create`, `set_openai_api_key`, and `mock_openai_client` functions to `test_utils.py`.

3. **Update `test_main.py`:**
   - Import the utility methods from `test_utils.py`.

### Implementation

Let's start by creating the `test_utils.py` file and updating the `test_main.py` file.

### New `test_utils.py`

```python
import pytest
from unittest.mock import patch, MagicMock
import os

def mock_openai_chat_completion_create(*args, **kwargs):
    class MockResponse:
        def __init__(self):
            self.choices = [self]

        @property
        def message(self):
            return {"content": "Mocked Ad Headline"}

    return MockResponse()

@pytest.fixture(autouse=True)
def set_openai_api_key():
    with patch.dict(os.environ, {"OPENAI_API_KEY": "test"}):
        yield

def mock_openai_client():
    mock_client = MagicMock()
    mock_client.chat.completions.create.side_effect = mock_openai_chat_completion_create
    return mock_client
```

### Updated `test_main.py`

```python
import pytest
from fastapi.testclient import TestClient
from unittest.mock import patch
import requests
from bs4 import BeautifulSoup
from app.main import app
from app.openai_utils import generate_headline, get_openai_client
from app.scraping import extract_images
from test_utils import set_openai_api_key, mock_openai_client

client = TestClient(app)

def test_generate_headline():
    text = "This is a sample website text."
    image_url = "http://example.com/image.jpg"
    with patch("app.openai_utils.get_openai_client", return_value=mock_openai_client()):
        headline = generate_headline(text, image_url)
        assert headline == "Mocked Ad Headline"

def test_extract_text_success():
    url = "http://example.com"
    html_content = """
    <html>
        <body>
            <p>Hello, World!</p>
            <img src="http://example.com/image1.jpg" />
            <img src="http://example.com/image2.jpg" />
        </body>
    </html>
    """
    expected_text = "Hello, World!"
    expected_images = ["http://example.com/image1.jpg", "http://example.com/image2.jpg"]
    expected_headlines = ["Mocked Ad Headline", "Mocked Ad Headline"]

    with patch("requests.get") as mock_get, patch("app.openai_utils.get_openai_client", return_value=mock_openai_client()):
        mock_get.return_value.status_code = 200
        mock_get.return_value.content = html_content

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert response.json() == {"text": expected_text, "images": expected_images, "headlines": expected_headlines}

def test_extract_text_invalid_url():
    response = client.get("/extract-text", params={"url": "invalid-url"})
    assert response.status_code == 422

def test_extract_text_request_exception():
    url = "http://example.com"

    with patch("requests.get") as mock_get:
        mock_get.side_effect = requests.RequestException("Request failed")

        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 400
        assert response.json() == {"detail": "Request failed"}

def test_extract_text_bermuda():
    url = "https://thinkingofbermuda.com"
    
    with patch("app.openai_utils.get_openai_client", return_value=mock_openai_client()):
        response = client.get("/extract-text", params={"url": url})
        assert response.status_code == 200
        assert "text" in response.json()
        assert "images" in response.json()
        assert "headlines" in response.json()

def test_extract_images():
    html_content = """
    <html>
        <body>
            <img src="http://example.com/image1.jpg" />
            <meta property="og:image" content="http://example.com/image2.jpg" />
            <link rel="image_src" href="http://example.com/image3.jpg" type="image/jpeg" />
            <source srcset="http://example.com/image4.jpg 1x, http://example.com/image5.jpg 2x" />
        </body>
    </html>
    """
    soup = BeautifulSoup(html_content, "html.parser")
    expected_images = [
        "http://example.com/image1.jpg",
        "http://example.com/image2.jpg",
        "http://example.com/image3.jpg",
        "http://example.com/image4.jpg"
    ]
    assert extract_images(soup) == expected_images
```

### Explanation of Changes

1. **New `test_utils.py`:**
   - Created a new file `test_utils.py` to hold the utility methods and fixtures.
   - Moved the `mock_openai_chat_completion_create`, `set_openai_api_key`, and `mock_openai_client` functions to `test_utils.py`.

2. **Updated `test_main.py`:**
   - Imported the utility methods from `test_utils.py`.

I will now update the PR with these changes.